### PR TITLE
Remove pkgs from nix profile that are no longer on devbox.json

### LINKED ISSUE #1138: [Bug]: When you manually remove a package from the config, devbox doesn't remove the package.
**Current Behavior (bug)**
Devbox will leave a package that has been manually removed from the config. So if I run `devbox update`, the package is still there even though it is not in `devbox.json`

**Expected Behavior (fix)**
When I remove a package from the config and run `devbox update` or `devbox install` the package should get removed. devbox should see that there is a package in the nix profile but is not in the config so it should remove it.

**Additional context**
Please include the output of `devbox version -v` and
a copy of your `devbox.json` file.

Info

```
Version:     0.5.4
Platform:    linux_amd64
Commit:      none
Commit Time: unknown
Go Version:  go1.20.4
Launcher:
```

JSON

```json
{
  "packages": [
    "hyfetch@latest",
    "htop@latest",
    "pfetch@latest",
    "neofetch@latest",
    "ranger@latest",
    "mc@latest",
    "gnome.gnome-tweaks@latest",
    "btop@latest",
    "vim@latest",
    "trash-cli@latest",
    "stow@latest",
    "kitty@latest",
    "mesa@latest"
  ],
  "shell": {
    "init_hook": [
      "echo 'Welcome to devbox!' > /dev/null",
      "export HOME=\"$HOME/custom-devbox/homedir\""
    ],
    "scripts": {
      "test": [
        "echo \"Error: no test specified\" && exit 1"
      ]
    }
  }
}
```

Nano still exists in the profile even though it is not there.

When I run `nix profile install nixpkgs#nano --profile ~/devboxdir/.devbox/nix/profile/default` and I run `devbox update` package doesn't get removed.

### PR DESCRIPTION
**Pull request by ipince**
## Summary

The goal is simple: if a package no longer exists in `devbox.json`, then it should be removed from the nix profile. The implementation is not as straightforward, because comparing packages can be very slow. Here we:

* Add memoization to `nix.Input` so that it remembers the normalized package attribute path if it ever had to resolve it.
* Compares list of of n packages in devbox.json against m pkgs in the profile, resulting in n+m calls, but _only_ if n != m, which is usually not the case in the happy path.
* Removes pkgs from the profile by passing in the indeces. This is so that we don't have to compare again whether an Input.Raw exists in the profile.

Notes/questions for future improvements:
* We're still creating a bunch of `nix.Input`s in a bunch of places, and we end up making more calls to get the normalized attribute path when comparing packages. We should add a global cache or something to avoid duplicate calls.
* Apart from avoiding _duplicate_ calls, I think there's ways to avoid the calls altogether:
  1. Could we change `devbox.lock` to already contain the normalized attribute paths in its `resolved` field? That way we can probably skip many of the normalization calls.
  2. When reading an item from the nix profile, can't we assume that the resolvedReference already has the normalized attribute path?

Fixes #1138 

## How was it tested?
Added a bunch of packages with `devbox add`. Removed them manually from `devbox.json` and called a devbox command. Added fine logging to inspect what was going on and ensure not too many extra expensive calls are being added. Example output below:
```
> nix profile list --profile .devbox/nix/profile/default
0 github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.go github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.go /nix/store/ihz3bv2mzrwxrjzq0i2ysin7kj4i61bv-go-1.20.3
1 github:NixOS/nixpkgs/242246ee1e58f54d2322227fc5eef53b4a616a31#legacyPackages.x86_64-darwin.actionlint github:NixOS/nixpkgs/242246ee1e58f54d2322227fc5eef53b4a616a31#legacyPackages.x86_64-darwin.actionlint /nix/store/68valrl03a40yzf3qcppfyq468cbc18k-actionlint-1.6.23
2 github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.golangci-lint github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.golangci-lint /nix/store/sc143w85a2sb6xxg6lp073az6rraa1s8-golangci-lint-1.52.2
3 github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.cowsay github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.cowsay /nix/store/c2a8wbqcss1y8yhnx5n1p3awn8cxw0pv-cowsay-3.7.0 /nix/store/v9dccca6yrg53j6zxa9wi46h5n5vxsbl-cowsay-3.7.0-man

> cat devbox.json
{
  "packages": [
    "go@1.20",
    "actionlint@1.6.23",
    "golangci-lint@1.52.2"
  ],
...

> db run -- cowsay 'hi' # after cowsay was manually removed
Ensuring packages are installed.

checking profile input (unresolved): github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.cowsay
  against: go@1.20 (resolved=github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#go)
    !! searched: legacyPackages.x86_64-darwin.cowsay
    !! searched: legacyPackages.x86_64-darwin.go
  against: actionlint@1.6.23 (resolved=github:NixOS/nixpkgs/242246ee1e58f54d2322227fc5eef53b4a616a31#actionlint)
  against: golangci-lint@1.52.2 (resolved=github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#golangci-lint)
    memoized: legacyPackages.x86_64-darwin.cowsay
    !! searched: legacyPackages.x86_64-darwin.golangci-lint

checking profile input (unresolved): github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.go
  against: go@1.20 (resolved=github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#go)
    !! searched: legacyPackages.x86_64-darwin.go
    memoized: legacyPackages.x86_64-darwin.go

checking profile input (unresolved): github:NixOS/nixpkgs/242246ee1e58f54d2322227fc5eef53b4a616a31#legacyPackages.x86_64-darwin.actionlint
  against: go@1.20 (resolved=github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#go)
  against: actionlint@1.6.23 (resolved=github:NixOS/nixpkgs/242246ee1e58f54d2322227fc5eef53b4a616a31#actionlint)
    !! searched: legacyPackages.x86_64-darwin.actionlint
    !! searched: legacyPackages.x86_64-darwin.actionlint

checking profile input (unresolved): github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.golangci-lint
  against: go@1.20 (resolved=github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#go)
    !! searched: legacyPackages.x86_64-darwin.golangci-lint
    memoized: legacyPackages.x86_64-darwin.go
  against: actionlint@1.6.23 (resolved=github:NixOS/nixpkgs/242246ee1e58f54d2322227fc5eef53b4a616a31#actionlint)
  against: golangci-lint@1.52.2 (resolved=github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#golangci-lint)
    memoized: legacyPackages.x86_64-darwin.golangci-lint
    memoized: legacyPackages.x86_64-darwin.golangci-lint
found 1 extras to remove
.cmd.sh: line 3: cowsay: command not found

> nix profile list --profile .devbox/nix/profile/default
0 github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.go github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.go /nix/store/ihz3bv2mzrwxrjzq0i2ysin7kj4i61bv-go-1.20.3
1 github:NixOS/nixpkgs/242246ee1e58f54d2322227fc5eef53b4a616a31#legacyPackages.x86_64-darwin.actionlint github:NixOS/nixpkgs/242246ee1e58f54d2322227fc5eef53b4a616a31#legacyPackages.x86_64-darwin.actionlint /nix/store/68valrl03a40yzf3qcppfyq468cbc18k-actionlint-1.6.23
2 github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.golangci-lint github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.golangci-lint /nix/store/sc143w85a2sb6xxg6lp073az6rraa1s8-golangci-lint-1.52.2
```

### CODE REVIEW

**Comment by mikeland86** [comment]
> Could we change devbox.lock to already contain the normalized attribute paths in its resolved field? That way we can probably skip many of the normalization calls.

We would need to save normalized names for devbox.json packages and also profile unlocked references.

I would suggest local.lock instead since it's purely a caching mechanism and not data that we care about.

File: internal/nix/profile.go [hunk 1]
```diff
@@ -264,6 +261,32 @@ func ProfileInstall(args *ProfileInstallArgs) error {
...
+	}
+
+	indexes := []string{}
+	for _, item := range items {
+		indexes = append(indexes, strconv.Itoa(item.index))
+	}
+	cmd := exec.Command("nix", "profile", "remove",
+		"--profile", profilePath,
+		"--impure", // for NIXPKGS_ALLOW_UNFREE
+		strings.Join(indexes, " "),
```
Comment (mikeland86) [inline internal/nix/profile.go:279]: Does this behave OK when multiple indexes? I think we can also remove by unlocked reference but that may be a problem with duplicates

**Review by mikeland86 — APPROVED** [review]
Nice! this LGTM. Just need to test perf edge cases. Especially flakes. You can try:

* Add many devbox packages (~20)
* Add flakes  (local and remote)
* Add php Haskel packages (that have plugins that create flakes) 
* Manually edit devbox.json to remove a few things. 
* Call something that triggers tidy and time it.

Also, take a look at these optimizations to make sure nothing regressed:
https://github.com/jetpack-io/devbox/pull/1050

File: internal/nix/input.go [hunk 2]
```diff
@@ -170,17 +176,33 @@ func (i *Input) FullPackageAttributePath() (string, error) {
...
-			return "", err
+	if i.IsDevboxPackage() {
+		if i.isVersioned() {
+			entry, err := i.lockfile.Resolve(i.Raw)
+			if err != nil {
+				return "", err
+			}
+			query = entry.Resolved
+		} else {
+			query = i.lockfile.LegacyNixpkgsPath(i.String())
```
[the remark this answers is not in the document]
Comment (ipince) [reply internal/nix/input.go:204]: Yes, this is a no-op change. I thought it was more clear this way.

It's just that `isVersioned` implies `isDevboxPkg`, but that's not immediately clear to a reader. And then the reader might think that the else-if branch applies to devbox packages. But it only really applies to non-versioned devbox packages. So, written this way makes the intention immediately clear without having to jump around the code.

But yeah, it's a bit subjective. I actually had also made the change on line 125 (that condition is always true).

[the remark this answers is not in the document]
Comment (ipince) [reply internal/nix/input.go:204]: I guess it's just weird/confusing to me to see control flow like:

```
if A && B {
  ...
} else if A {
  ...
} else {
  ...
}
```

File: internal/nix/input.go [hunk 3]
```diff
@@ -24,6 +24,8 @@ type Input struct {
 	url.URL
 	lockfile lock.Locker
 	Raw      string
+
+	normalizedAttributePath string // memoized value from normalizedPackageAttributePath()
```
[the remark this answers is not in the document]
Comment (ipince) [reply internal/nix/input.go:28]: Will do

File: internal/impl/packages.go [hunk 4]
```diff
@@ -219,6 +216,16 @@ func (d *Devbox) profilePath() (string, error) {
 	return absPath, errors.WithStack(os.MkdirAll(filepath.Dir(absPath), 0755))
 }
 
+func (d *Devbox) syncPackagesToProfile(ctx context.Context, mode installMode) error {
```
[the remark this answers is not in the document]
Comment (ipince) [reply internal/impl/packages.go:219]: Will do

**Review by ipince — COMMENTED** [review]
> Add php Haskel packages (that have plugins that create flakes)

Do you have an example? If not, I'll use `haskellPackages.acme-php`.

Sidenote: I tried searching for that using devbox search but failed, probably because we only allow single-word searches. Would be nice to allow multiple words in the future.

**Comment by ipince** [comment]
> We would need to save normalized names for devbox.json packages and also profile unlocked references.

Hmm, why?

Anyway, my original suggestion was to save the normalized package, like:
```
github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#legacyPackages.x86_64-darwin.go
```

instead of what we do today, which is this:
```
github:NixOS/nixpkgs/8670e496ffd093b60e74e7fa53526aa5920d09eb#go
```

But I guess that won't really work, because the normalized attribute paths are platform-dependent. We could save all of them, sure, but at that point it's probably not worth it to save a call to `nix search`.

--

My other suggestion was that, when we convert a `ProfileItem` to an `Input`, we can basically do `input.normalizedAttributeCache = item.lockedReference`. They should _always_ be identical, right?

File: internal/impl/packages.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ipince) [reply internal/impl/packages.go:219]: Done

File: internal/nix/input.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ipince) [reply internal/nix/input.go:28]: Done. renamed to `normalizedPackageAttributePathCache`. IT's long af but consistent with the other PackageAttributePath. I think we should drop the "Package" everywhere, but I didn't want to add scope to this PR. We can change that separately.

**Comment by ipince** [comment]
OK I finished testing the performance stuff, and also local and remote flakes.

Performance-wise, the comparisons are fairly quick. It at most took ~4 seconds with doing it with ~20 packages. Since this scenario should not happen very often, I think it's fine to ship.

I did notice that sometimes there's a lot of time spent _after_ the profile syncing, and I'm not sure what is it, but it's unrelated and should be part of our efforts to improve performance overall. I think there's some relatively low-hanging fruit.

Merging after tests pass.
